rust/otap-dataflow/crates/otlp/src/batch_processor.rs

```diff
@@ -0,0 +1,248 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! BatchProcessor: Buffers messages and emits them in batches.
+use async_trait::async_trait;
+use otap_df_engine::error::Error;
+use otap_df_engine::message::{ControlMsg, Message};
+use otap_df_engine::processor::{SendEffectHandler, Processor, EffectHandlerTrait};
+use std::time::Duration;
+
+/// A processor that buffers messages and emits them in batches.
+pub struct BatchProcessor<PData> {
+    batch: Vec<PData>,
+    batch_size: usize,
+}
+
+#[allow(dead_code)]
+impl<PData> BatchProcessor<PData> {
+    pub fn new(batch_size: usize) -> Self {
+        Self {
+            batch: Vec::with_capacity(batch_size),
+            batch_size,
+        }
+    }
+}
+
+#[async_trait(?Send)]
+impl<PData: Clone + Send + 'static> Processor<PData, SendEffectHandler<PData>> for BatchProcessor<PData> {
+
+    async fn process(
+        &mut self,
+        msg: Message<PData>,
+        effect_handler: &mut SendEffectHandler<PData>,
+    ) -> Result<(), Error<PData>> {
+        match msg {
+            Message::PData(data) => {
+                self.batch.push(data);
+                if self.batch.len() >= self.batch_size {
+                    let batch = std::mem::take(&mut self.batch);
```

Comment:
(I know its draft, but still leaving a question).. When `take` is invoked, that self.batch will now point to a default - i.e empty vec! - is that the intention?

Reply:
yes, the `take` resets the batch and does not create a new 'vec', however, now I wonder if we should be able to call an `EffectHandler` API that handles the `vec` of `pdata` items? Perhaps it does not matter as at some point each data item will need to be read and dealt with?

Reply:
I was mainly trying to ask a different question (I now realized I did a poor job framing it initially).

Each time take (or replace) is used, we are going to get a *new* vector. Could that result in creating new and new vec! unnecessarily throughout the lifetime? Could we instead create two vec! of desired capacity during initialization, and keep swapping them when batch_size reaches the limit, so utmost we'll only have 2 vec! and that'll be reused throughout.

(Disclaimer: I am new to the repo, so I might not have fully understood the code)

Reply:
According to my reading, `replace` is more efficient here. I am unsure about using 2 and swapping  - does this avoid any new allocations at all? Is it possible that approach will be more complex than needed?